product_variant_default_code/models/product.py

```diff
@@ -186,18 +158,43 @@ def _compute_default_code(self):
 class ProductProduct(models.Model):
     _inherit = "product.product"
 
-    manual_code = fields.Boolean(string="Manual Reference")
-
-    @api.model
-    def create(self, vals):
-        product = super(ProductProduct, self).create(vals)
-        if product.reference_mask:
-            render_default_code(product, product.reference_mask)
-        return product
+    manual_code = fields.Boolean(string="Manual Reference", default=False)
+    default_code = fields.Char(
+        compute="_compute_default_code", readonly=False, store=True
+    )
 
-    @api.onchange("default_code")
-    def onchange_default_code(self):
-        self.manual_code = bool(self.default_code)
+    @api.depends(
+        "product_tmpl_id.reference_mask",
+        "product_template_attribute_value_ids.attribute_id.code",
+        "product_template_attribute_value_ids.attribute_id.name",
+        "product_template_attribute_value_ids.attribute_id.sequence",
+        "product_template_attribute_value_ids.product_attribute_value_id.code",
+        "product_template_attribute_value_ids.product_attribute_value_id.name",
+    )
+    def _compute_default_code(self):
+        for rec in self:
+            if not rec.manual_code:
+                product_attrs = defaultdict(str)
+                reference_mask = ReferenceMask(rec.product_tmpl_id.reference_mask)
+                main_lang = rec.product_tmpl_id._guess_main_lang()
+                for attr in rec.product_template_attribute_value_ids:
+                    value = attr.product_attribute_value_id
+                    attr_name = value.attribute_id.with_context(lang=main_lang).name
+                    if value.attribute_id.code:
+                        product_attrs[attr_name] += value.attribute_id.code
+                    if value.code:
+                        product_attrs[attr_name] += value.code
+                all_attrs = extract_token(rec.product_tmpl_id.reference_mask)
+                missing_attrs = all_attrs - set(product_attrs.keys())
+                missing = dict.fromkeys(
+                    missing_attrs,
+                    rec.env["ir.config_parameter"]
+                    .sudo()
+                    .get_param("default_reference_missing_placeholder"),
+                )
+                product_attrs.update(missing)
+                default_code = reference_mask.safe_substitute(product_attrs)
```

Comment:
can you move this code in a method _build_default_code or _generate_default_code.
But this method is normal odoo method of the class product.product (not like before with the method get_rendered_default_code)